Move pthread_tryjoin_np to native code. NFC

sbc100 opened the pull request:
This matches more closely the musl version.

kripken commented on system/lib/pthread/pthread_create.c:
```diff
@@ -155,6 +155,8 @@ void _emscripten_thread_exit(void* result) {
   // Not hosting a pthread anymore in this worker set __pthread_self to NULL
   _emscripten_thread_init(0, 0, 0);
+  self->tid = 0;
+  self->dead = 1;
```
grepping `system/` I can't find any `.dead` or `->dead`. How is it used?

kripken replied:
Ah, looks like you removed this line meanwhile as I was reviewing...

sbc100 replied:
Sorry, this was from a previous version of the PR..  dead was removing the v1.2.2 update that just landed and then I rebased.

kripken commented on system/lib/pthread/pthread_join.c:
```diff
@@ -9,8 +9,31 @@
 extern int __pthread_join_js(pthread_t thread, void **retval);
+
 int __pthread_join(pthread_t thread, void **retval) {
+  if (thread->self != thread) {
```
The JS version checked if `thread` was NULL before doing these, do we not still want that?

sbc100 replied:
None of the musl entry points to this.. so it seems unnecessary.. calling these functions with NULL in undefined behaviour,  I suppose we could at a check in debug builds, but we might also need to then add a debug configuration of libc.   I'm tempted to go with musl on this one.

kripken replied:
Fair enough. ASan should catch bugs there anyhow.

kripken commented on system/lib/pthread/pthread_join.c:
```diff
@@ -9,8 +9,30 @@
+    // not exist anymore.
+    return ESRCH;
+  }
+  if (thread->detach_state == DT_DETACHED) {
```
Do we not need to access this atomically? JS seemed to (on `detached`, which I can't seem to find.) It seems to be defined as "volatile", not atomic, in pthread_impl.h

sbc100 replied:
Yes, musl is using volatile here for these locations rather than atomics.   This is what they do for all their atomics api in `src/internal/atomic.h`.

I think this pre-check is equivalent to the existing one.  We still need to handle the case where the thread becomes detached *while* are joining it.. and I don't thing the JS code currently does that.. but that is a separate issue I think.

sbc100 replied:
I changed this to use `emscripten_atomic_load_u32` to match the existing JS impl.    And I added TODO.

sbc100 replied:
having made that change I'm not sure its really necessary.  

musl code internally seems perfectly happy to read this variable without any special atomic functions:

https://github.com/emscripten-core/musl/blob/85e0e3519655220688e757b9d5bfd314923548bd/src/thread/pthread_join.c#L16

and:

https://github.com/emscripten-core/musl/blob/85e0e3519655220688e757b9d5bfd314923548bd/src/thread/pthread_join.c#L35

In these cases I think as long as we don't get tearing then its fine to read out-of-date values (and race with other threads).

see

kleisauke reviewed (approved):
LGTM! Though, I wonder about that `posixtest.test_pthread_detach_1_2` test failure (IIRC, the same issue is also present in #13007 when I moved the whole `pthread_join` to native code, it was the only test I had to [temporarily disable](https://github.com/kleisauke/emscripten/blob/musl-1.2.1-merge/tests/test_posixtest.py#L132)).

sbc100 commented:
> LGTM! Though, I wonder about that `posixtest.test_pthread_detach_1_2` test failure (IIRC, the same issue is also present in #13007 when I moved the whole `pthread_join` to native code, it was the only test I had to [temporarily disable](https://github.com/kleisauke/emscripten/blob/musl-1.2.1-merge/tests/test_posixtest.py#L132)).

I found and fixed that issue.

sbc100 commented:
I made a few final changes to this CL (mostly adding the extra `tryjoin` argument back), so maybe take another look post-landing..

kripken commented:
Do you have a diff of those final changes? Adding more commits to PRs instead of constantly rebasing into a single commit would have made that easier to figure out. I do think that is a good model to follow in general as I've said.

sbc100 commented:
Sorry github doesn't keep track of that information very well it seems.     I wish there was an easy way in the UI to show the differences. 

I didn't realized you were suggesting that I avoid squashing in general went we chatting about this on the musl update PR.  I can try to move in that directions but it will require some changes to my workflow.  I'm very much about keeping each PR in a single commit and using `git show` to see the current state and `git commit --amend` to update it.  I will create an alias to replace `git show`.. something like `git diff @{upstream}...` should do it I think.

kripken commented:
Ah, I see. Well, I don't want to disrupt your general workflow, in specific PRs where this comes up we can figure something out I guess.

For this particular PR, do you recall what those last changes were?

kripken commented:
Skimming the code now it looks reasonable, but I didn't read it as thoroughly as before where I compared each line to the previous JS  for it etc.

sbc100 commented:
> Ah, I see. Well, I don't want to disrupt your general workflow, in specific PRs where this comes up we can figure something out I guess.
> 
> For this particular PR, do you recall what those last changes were?

For this PR, the final change was to re-introduce the third argument to pthread_join_js so the JS could know if was being called from `tryjoin` or regular `join`.. the only difference is that `tryjoin` will not warn about blocking.

kripken commented:
I see, thanks. That change lgtm.
